src/ProjectOrigin.WalletSystem.Server/Repositories/CertificateRepository.cs

```diff
@@ -100,53 +93,119 @@ public async Task InsertCertificate(Certificate certificate)
+        {
+            var totalCouunt = gridReader.ReadSingle<int>();
+            var certificates = gridReader.Read<CertificateViewModel>();
+            var attributes = gridReader.Read<AttributeViewModel>();
 
-        await _connection.QueryAsync<CertificateViewModel, SliceViewModel, CertificateAttribute, CertificateViewModel>(
-            @"SELECT c.*, s.Id AS slice_id, s.quantity, a.attribute_key as key, a.attribute_value as value, a.attribute_type as type
-              FROM Wallets w
-              INNER JOIN wallet_endpoints re
-                ON w.id = re.wallet_id
-              INNER JOIN wallet_slices s
-                ON re.Id = s.wallet_endpoint_id
-              INNER JOIN certificates c
-                ON s.certificate_id = c.id
-              LEFT JOIN attributes_view a
-                ON c.id = a.certificate_id
-                AND c.registry_name = a.registry_name
-                AND (a.wallet_id = w.id OR a.wallet_id IS NULL)
-              WHERE w.owner = @owner
-                AND s.state = @sliceState
-                AND (@start IS NULL OR c.start_date >= @start)
-                AND (@end IS NULL OR c.end_date <= @end)
-                AND (@type IS NULL OR c.certificate_type = @type)
-                ",
-            (cert, slice, atr) =>
+            foreach (var certificate in certificates)
             {
-                if (!certsDictionary.TryGetValue(cert.Id, out var certificate))
-                {
-                    certificate = cert;
-                    certsDictionary.Add(cert.Id, cert);
-                }
-                if (slice != null && !certificate.Slices.Contains(slice))
-                    certificate.Slices.Add(slice);
-                if (atr != null && atr.Key != null && atr.Value != null && !certificate.Attributes.Contains(atr))
-                    certificate.Attributes.Add(atr);
-                return certificate;
-            },
-            splitOn: "slice_id, key",
-            param: new
+                certificate.Attributes.AddRange(attributes
+                    .Where(attr => attr.RegistryName == certificate.RegistryName
+                            && attr.CertificateId == certificate.CertificateId));
+            }
+
+            return new PageResult<CertificateViewModel>()
             {
-                owner,
-                sliceState = (int)WalletSliceState.Available,
-                start = filter.Start,
-                end = filter.End,
-                type = filter.Type,
-            });
+                Items = certificates,
+                TotalCount = totalCouunt,
+                Count = certificates.Count(),
+                Offset = filter.Skip,
+                Limit = filter.Limit
+            };
+        }
+    }
 
-        return certsDictionary.Values;
+    public async Task<PageResult<AggregatedCertificatesViewModel>> QueryAggregatedAvailableCertificates(CertificatesFilter filter, TimeAggregate timeAggregate, string timeZone)
```

Comment:
Why is this a new function? Didn't we have this already?

Reply:
aggregation was previously done in the service, now it is done on database level.